### DISCUSSION TITLE
[rejected] disputes.messages.author.role

### DISCUSSION THREAD

**Discussion by maniootek** [allegro/allegro-api#5630] [Suggestions - rejected]
Zgodnie ze specyfikacją na stronie:
https://developer.allegro.pl/tutorials/jak-zarzadzac-dyskusjami-E7Zj6gK7ysE#kliknij-aby-zobaczyc-response
są dwie możliwe opcje: BUYER, ADMIN
a nie powinno być jeszcze dopisane "SYSTEM" i "SELLER" ?

**Comment by maniootek** [comment]
Chodzi o to, że szukam sposobu na napisanie autorespondera. Chciałbym ustalić założenia. Póki co mam takie:
1. pobieram dyskusje
2. jeżeli "status" danej dyskusji jest inny niż "ONGOING" to pomijam dyskusje idę do następnej
3. pobieram wiadomości dla danej dyskusji i sprawdzam je od najnowszej
4. jeżeli "author.role" dla ostatniej wiadomości jest "SELLER" to idę do następnej dyskusji
5. jeżeli "author.role" dla ostatniej wiadomości jest "BUYER" to odpisuję na dyskusję i idę do następnej dyskusji
6. jeżeli "author.role" dla ostatniej wiadomości jest inne niż "SELLER" i inne niż "BUYER" to idę do następnej wiadomości dla danej dyskusji

Czy dobrze to sobie ustaliłem?

**Comment by Lukasz-Zurek** [comment]
@maniootek Opcje **"BUYER"** i **"ADMIN"** w polu **"author.role"** oznaczają kto założył dyskusję. W tym przypadku rozpocząć ją może Kupujący, lub w wyjątkowych przypadkach - administrator/pracownik Allegro.pl.

Skorzystaj z innego pola - **"messagesStatus"**, w którym zwracamy następujące wartości:

**"NEW"** - nowa dyskusja (w której sprzedający jeszcze nie odpowiedział),
**"BUYER_REPLIED"** - nowa wiadomość od kupującego,
**"SELLER_REPLIED"** - nowa wiadomość od sprzedającego,
**"ALLEGRO_ADVISOR_REPLIED"** - nowa wiadomość od Allegro.

Myślę, że dzięki temu łatwiej ustawisz autorespondera.

**Reply by maniootek** [reply]
Nie rozumiem. Jak to pole **author.role** oznacza kto założył dyskusję? Przecież to pole jest przypisane do zbioru **message** i oznacza kto napisał daną wiadomość. Czyż nie?

**Comment by Lukasz-Zurek** [comment]
@maniootek Tak, jednak obiekt **"message"** oznacza początkową wiadomość w dyskusji (DisputeFirstMessage) natomiast **"author"** to twórca dyskusji (DisputeAuthor). Jeżeli przykładowo Kupujący tworzy dyskusję - musi wprowadzić treść i ewentualnie załączyć plik. Jest to treść początkowa w dyskusji/pierwsza wiadomość. 

Opisy pól sprawdzisz w naszej [dokumentacji](https://developer.allegro.pl/documentation/#operation/getListOfDisputesUsingGET).

**Comment by maniootek** [comment]
Tylko mi chodzi o obiekt message z zasobu messages a nie z zasobu disputes

**Reply by Lukasz-Zurek** [reply]
@maniootek Twój wątek wyraźnie dotyczy dyskusji.

**Comment by maniootek** [comment]
Ja do tej pory najpierw pobierałem disputes a potem messages do danego disputa i potem analizowałem wszystkie wiadomości ale chyba niepotrzebnie. Samo pobieranie disputes chyba wystarczy.

**Comment by maniootek** [comment]
Jeżeli dobrze to rozszyfrowałem, to żeby stworzyć auto-respondera to pobieramy disputes i sprawdzamy tylko messageStatus danego disputa i jeżeli wartość tego pola to NEW to wtedy wysyłamy automatyczną odpowiedź. Zgadza się?

**Reply by maniootek** [reply]
NEW i BUYER_REPLIED

**Reply by maniootek** [reply]
jeszcze wcześniej trzeba sprawdzić status i odpowiadać tylko dla tych disputów, które w tym polu mają wartość ONGOING

**Reply by Lukasz-Zurek** [reply]
Dokładnie tak, **"messagesStatus"**: **"NEW"** i **"BUYER_REPLIED"** +  **"disputes.status"**: **"ONGOING"**. Pozostaje jeszcze **"messagesStatus"**: **"ALLEGRO_ADVISOR_REPLIED"**, ale tutaj pracownik Allegro.pl najczęściej prosi o wyjaśnienia, lub konkretne informacje/dane, lub załączniki, więc w przypadku autorespondera można to zignorować, bo i tak Sprzedający musi się tutaj sam wypowiedzieć.

**Reply by maniootek** [reply]
Do tego jeszcze wymyśliłem, żeby sprawdzać **lastMessageCreationDate** i odpowiadać dopiero po **15min** od tej daty. Chodzi o to, że daje trochę czasu dla pracownika, aby to jednak on odpisał a nie automat oraz chcę uniknąć sytuacji, kiedy to autoresponder wcina się między wymianę wiadomości między pracownikiem a kupującym.

**Reply by maniootek** [reply]
jak się nazywa taki format zapisu daty i czasu? przykład: **2018-09-14T06:34:24.258Z**

**Reply by Lukasz-Zurek** [reply]
Wszystkie daty (m.in. w zasobie dotyczącym dyskusji) prezentowane są w formacie ISO 8601. Więcej informacji [tutaj](https://developer.allegro.pl/tutorials/informacje-podstawowe-b21569boAI1#data-i-czas).

**Reply by Lukasz-Zurek** [reply]
> Do tego jeszcze wymyśliłem, żeby sprawdzać **lastMessageCreationDate** i odpowiadać dopiero po **15min** od tej daty. Chodzi o to, że daje trochę czasu dla pracownika, aby to jednak on odpisał a nie automat oraz chcę uniknąć sytuacji, kiedy to autoresponder wcina się między wymianę wiadomości między pracownikiem a kupującym.

Pracownik Allegro.pl włączy się do Dyskusji w sytuacji, gdy sprzedający nie udzielił odpowiedzi w wymaganym czasie (24 godziny), lub jeśli Kupujący poprosi go o wsparcie i tu czas może być różny. Czas 15 minut może okazać się wystarczający, ale z pewnością nie dla wszystkich przypadków. Możesz go oczywiście wydłużyć. Aczkolwiek trudno znaleźć tutaj idealne rozwiązanie.

**Comment by maniootek** [comment]
Coś jest nie tak, mam dyskusję gdzie to ja odpowiedziałem ostatni (auto-responderem) ale status jest BUYER_REPLIED i lastMessageCreationDate wskazuje na czas odpowiedzi kupującego a nie mojej wiadomości

**Reply by Lukasz-Zurek** [reply]
Podaj proszę id dyskusji i daj proszę znać czy chodzi o sandbox czy produkcję. 

Dziękuję.

**Reply by maniootek** [reply]
nie umiem już tej sytuacji odtworzyć, na razie jest ok, bardzo dziękuje za informacje

**Reply by maniootek** [reply]
jednak udało się, dispute_id: 49423ed7-90ac-4520-8719-629e84d5c22c
przed chwilą tam odpisał autoresponder a api zwraca dalej MessageStatus jako NEW

**Reply by maniootek** [reply]
api produkcyjne

**Reply by maniootek** [reply]
po chwili już zmieniło status ale tak jakby api nie odświeża danych zaraz po wysłaniu wiadomości i mój autoresponder wysyła podwójnie wiadomości

**Reply by maniootek** [reply]
czasami czas odświeżenia jest krótki a czasami długi

**Reply by Lukasz-Zurek** [reply]
Nie odnotowałem tutaj opóźnień po naszej stronie.

**Pierwsza wiadomość autorespondera:** 2022-05-27 08:57:35.932
**Druga wiadomość autorespondera:** 2022-05-27 08:57:35.951
**Odświeżenie statusu:** 2022-05-27 08:57:35.973

Faktycznie widać tutaj, że odświeżenie statusu nastąpiło po wysłaniu drugiego autorespondera. Zalecam modyfikację po Twojej stronie, np. dodanie warunku wysyłania tylko jednego autorespondera podczas statusu "NEW" (do momentu zmiany statusu), lub wysłanie autorespondera w danym określonym czasie, przy uwzględnieniu statusu "NEW". Trudno mi tutaj coś odpowiedniego zasugerować. Mam nadzieję, że uda się Tobie dostosować autoresponder, aby funkcjonował prawidłowo.

**Reply by maniootek** [reply]
Czy na pewno została przeanalizowana prawidłowa dyskusja? Tam nie było drugiej wiadomości autorespondera tylko jedna. Kolejna wiadomość z naszej strony miała miejsce po 2h ale już ręcznie przez pracownika (przez panel webowy)
![ar](https://user-images.githubusercontent.com/45625552/170669026-ca34c604-e93a-4491-9348-2178bd5c6362.png)

**Reply by maniootek** [reply]
W tej dyskusji było tak, że jak autoresponder wysłał wiadomość, to po kilku/kilkunastu sekundach pobrałem ten dispute prze API i jego status dalej był NEW. Potem zrobiłem to samo i już status się zmienił. Tak jakby allegro API potrzebowało chwili, żeby ten status zaktualizować.

**Reply by Lukasz-Zurek** [reply]
Faktycznie tak było. Kwestię dotyczącą odświeżenia statusu przekażę do weryfikacji.

**Comment by maniootek** [comment]
Jeżeli jest taka możliwość to prosił bym jeszcze o sprawdzenie dyskusji o identyfikatorze 28dd1e76-1e5e-4792-88bb-20ec685f0561, tam po 10 minutach od napisania wiadomości auto-responderem, program wysłał ponownie wiadomość a nie powinien, czyli tak jakby status odświeżył się znacznie później (pozostał jako NEW nawet po 10 minutach). Niestety nie mam logów do aplikacji i nie wiem jak to dokładnie wyglądało. Logi włączyłem dzisiaj i będę je analizował jeżeli podobna sytuacja się pojawi i wtedy też wrócę tutaj z aktualizacją tematu.

**Reply by Lukasz-Zurek** [reply]
Sprawę dotyczącą tej dyskusji również przekazałem do weryfikacji.

**Reply by Lukasz-Zurek** [reply]
Generalnie status ustalamy po odpytaniu dyskusji za pomocą GET /sale/disputes/{disputeId}. Czyli nie ma tu żadnych opóźnień. Po prostu "messageStatus" zmieni się w momencie, gdy skorzystasz z powyższego endpointu (po wysłaniu autorespondera).

**Reply by maniootek** [reply]
ale ja dokładnie w taki sposób to sprawdzam

**Reply by maniootek** [reply]
czy w tej dyskusji 28dd1e76-1e5e-4792-88bb-20ec685f0561 czas zmiany statusu był prawidłowy czy opóźniony?

**Reply by Lukasz-Zurek** [reply]
Dla tej dyskusji pomiędzy pierwszą a drugą wiadomością nie odnotowałem odpytania dyskusji za pomocą GET /sale/disputes/{disputeId}. Jeśli było inaczej - podeślij proszę trace-id.

**Reply by maniootek** [reply]
ja pobieram dyskuje w ten sposób:
Get Request: https://api.allegro.pl/sale/disputes?limit=100&offset=0

**Reply by Lukasz-Zurek** [reply]
Rozumiem. Takiego requestu nie odnotowałem jednak w dniu 27.05, w przedziale czasowym pomiędzy 7:35 a 7:45. Z tego też powodu status się nie zmienił.

**Reply by maniootek** [reply]
Czyli pobieranie dyskusji za pomocą GET /sale/disputes?limit=100&offset=0 może powodować pobranie dyskusji z nieaktualnym statusem i muszę po wysłaniu wiadomości za pomocą POST /sale/disputes/{disputeID}/messages pobrać dodatkowo daną dyskusję za pomocą GET /sale/disputes/{disputeId} w celu zaktualizowania jego statusu, zgadza się?

**Reply by Lukasz-Zurek** [reply]
Tak. Musisz pobrać dodatkowo daną dyskusję za pomocą GET /sale/disputes/(disputeId} w celu zaktualizowania statusu.

**Reply by maniootek** [reply]
To trochę dziwne, że GET /sale/disputes?limit=100&offset=0 tego nie robi

**Comment by maniootek** [comment]
Wracam z aktualizacją ponieważ przeanalizowałem logi po swojej stronie. Chodzi o dyskusję id 34a1c8eb-c285-4b0e-9c78-672d536c6715. Wczytałem ją za pomocą GET /sale/disputes?limit=100&offset=0 (30-05-2022 18:45:17) i wynik wskazał "messagesStatus":"NEW" dlatego wysłałem za pomocą POST /sale/disputes/34a1c8eb-c285-4b0e-9c78-672d536c6715/messages nową wiadomość (auto-responder). Potem wykonałem polecenie GET /sale/disputes/34a1c8eb-c285-4b0e-9c78-672d536c6715 bo wg naszej wcześniejszej korespondencji wywnioskowałem, że to polecenie aktualizuje w bazie status. Faktycznie, polecenie zwróciło "messagesStatus":"SELLER_REPLIED" (czyli prawidłowo) ALE jak chwile później znowu wykonałem polecenie GET /sale/disputes?limit=100&offset=0 (30-05-2022 18:50:54) to "messagesStatus" ma wartość "NEW". To wszystko wygląda jakby polecenie GET /sale/disputes?limit=100&offset=0 nie pobierało aktualnego statusu dyskusji. Mogę przesłać loga jak coś

**Reply by Lukasz-Zurek** [reply]
Bardzo Ci dziękuję za dokładny opis i przykłady. Zgłaszam to do odpowiedniego zespołu. Wkrótce wrócę z informacją.

**Reply by maniootek** [reply]
Warto dodać, że taka sytuacja nie dzieje się z każdą dyskusją. Np. od wczoraj wieczora do dzisiaj rana była tylko jedna taka sytuacja.

**Reply by maniootek** [reply]
Kolejny przykład to dyskuja o id e18c9f1e-911b-4c8a-bb7b-7ef795cb0d72

**Reply by Lukasz-Zurek** [reply]
Dziękuję za kolejny przykład. Mogę już potwierdzić, że jest problem po naszej stronie. Obecnie zespół odpowiedzialny za dyskusje pracuje nad wyeliminowaniem tych utrudnień.

**Reply by maniootek** [reply]
Już naprawione? Bo nie przytrafiła mi się już zdublowana wiadomość w ostatnich dniach.

**Reply by Lukasz-Zurek** [reply]
Jest to problem "globalny", wymagający zmiany podejścia po naszej stronie. Rozwiązanie tych utrudnień będzie wymagało czasu. Proszę o cierpliwość.

**Reply by Lukasz-Zurek** [reply]
@maniootek Przepraszam Cię za długi czas oczekiwania na odpowiedź. Po naszej stronie przygotowaliśmy zmiany, dzięki którym nie powinno już dojść do podobnej sytuacji w trwających dyskusjach.

Raz jeszcze przepraszam za dotychczasowe utrudnienia, które komplikowały pracę nad autoresponderem.

**Reply by maniootek** [reply]
Bardzo dziękuję za reakcje. Problem już nie występuje.
